consensus/hotstuff/votecollector/combined_vote_processor_v2.go

```diff
@@ -0,0 +1,232 @@
+				vote.ID(), vote.View, err)
+		}
+		return fmt.Errorf("internal error checking signature validity for vote %v: %w", vote.ID(), err)
+	}
+
+	if p.done.Load() {
+		return nil
+	}
+
+	// verify random beacon sig
+	if randomBeaconSig != nil {
+		err = p.rbRector.Verify(vote.SignerID, randomBeaconSig)
+		if err != nil {
+			if errors.Is(err, msig.ErrInvalidFormat) {
+				return model.NewInvalidVoteErrorf(vote, "vote %x for view %d has an invalid random beacon signature: %w",
+					vote.ID(), vote.View, err)
+			}
+			return fmt.Errorf("internal error checking signature validity for vote %v: %w", vote.ID(), err)
+		}
+	}
+
+	if p.done.Load() {
+		return nil
+	}
+
+	// aggregate staking sig
+	_, err = p.stakingSigAggtor.TrustedAdd(vote.SignerID, stakingSig)
+	if err != nil {
+		return fmt.Errorf("adding the signature to staking aggregator failed for vote %v: %w", vote.ID(), err)
+	}
+	// aggregate random beacon sig
+	if randomBeaconSig != nil {
+		_, err = p.rbRector.TrustedAdd(vote.SignerID, randomBeaconSig)
+		if err != nil {
+			return fmt.Errorf("adding the signature to random beacon reconstructor failed for vote %v: %w", vote.ID(), err)
+		}
+	}
+
+	// checking of conditions for building QC are satisfied
+	if p.stakingSigAggtor.TotalWeight() < p.minRequiredStake {
+		return nil
+	}
+	if !p.rbRector.HasSufficientShares() {
+		return nil
+	}
+
+	// At this point, we have enough signatures to build a QC. Another routine
+	// might just be at this point. To avoid duplicate work, only one routine can pass:
+	if !p.done.CAS(false, true) {
+		return nil
+	}
+
+	// Our algorithm for checking votes and adding them to the aggregators should
+	// guarantee that we are _always_ able to successfully construct a QC when we
+	// reach this point. A failure implies that the VoteProcessor's internal state is corrupted.
+	qc, err := p.buildQC()
+	if err != nil {
+		return fmt.Errorf("internal error constructing QC from votes: %w", err)
+	}
+
+	p.onQCCreated(qc)
+
+	return nil
+}
+
+// buildQC performs aggregation and reconstruction of signatures when we have collected enough
+// signatures for building a QC. This function is run only once by a single worker.
+// Any error should be treated as exception.
+func (p *CombinedVoteProcessorV2) buildQC() (*flow.QuorumCertificate, error) {
+	stakingSigners, aggregatedStakingSig, err := p.stakingSigAggtor.Aggregate()
+	if err != nil {
+		return nil, fmt.Errorf("could not aggregate staking signature: %w", err)
+	}
+	reconstructedBeaconSig, err := p.rbRector.Reconstruct()
+	if err != nil {
+		return nil, fmt.Errorf("could not reconstruct random beacon group signature: %w", err)
+	}
+
+	blockSigData := &hotstuff.BlockSignatureData{
+		StakingSigners:               stakingSigners,
+		RandomBeaconSigners:          nil,
+		AggregatedStakingSig:         aggregatedStakingSig,
+		AggregatedRandomBeaconSig:    nil,
+		ReconstructedRandomBeaconSig: reconstructedBeaconSig,
+	}
```

Comment:
I would suggest to remove the unused fields

```suggestion
	blockSigData := &hotstuff.BlockSignatureData{
		StakingSigners:               stakingSigners,
		AggregatedStakingSig:         aggregatedStakingSig,
		ReconstructedRandomBeaconSig: reconstructedBeaconSig,
	}
```